### ISSUE TITLE
Add Dependencies incrementally on the fly from REPL..?

### ISSUE THREAD

**Issue by azizzaeny** [boot-clj/boot#499]
while compile build running from background, is it be possible OR could we establish some pattern? to pull dependencies to local classpath..? 

First attemp: 
-  create folder resource, src and build.boot

``` clojure
(set-env!
  :resource-paths #{"resources","src"}
  :dependencies '[
                  [org.clojure/clojurescript "1.9.89"]
                  [adzerk/boot-reload "0.4.11" :scope "test"]
                  [pandeiro/boot-http "0.7.3" :scope "test"]
                  [adzerk/boot-cljs "1.7.170-3"]])

(require '[pandeiro.boot-http :refer [serve]]
             '[adzerk.boot-cljs :refer [cljs]]
             '[adzerk.boot-reload :refer [reload]])

(def c (cljs)) ; cache deref
```
- Second add first namespace to source `src/app/core.cljs` and `boot repl` from terminal.  

``` clojure
(ns app.core)
(js/console.log "yap we are connected")
```

``` clojure
boot.user => (def f (future (boot (watch) c)))
Starting file watcher (CTRL-C to quit)...
Elapsed time: 21.499 sec
```
- Couple minute letter...i decided to use quiescient and sablono,..
  lets add it to our boot-env like we have done before...

``` clojure
boot.user => (set-env! :dependencies ........ [quiescent "0.3.2"] [sablono "0.7.4"])
```

because we dont have sablono from local maven repo its seem pulling from clojars....
- Then edit  src/app/core.cljs to:

``` clojure
(ns app.core)
  (:require [sablono.core :as html]
                [quiescent.core :as q])) 

(js/console.log "yap we are connected")
```

then i noticed that ..still boot pulling new jar file to m2 repo, its not add to classpath current running boot-env.. :

``` sh
Compiling ClojureScript...
• main.js
 No such namespace: quiescent.core, could not locate quiescent/core.cljs,...
 No such namespace: sablono.core...
```

refering issue cljs : #[boot-cljs/95](https://github.com/adzerk-oss/boot-cljs/issues/95)

**Comment by Deraen** [comment]
Boot does support this.

You should use e.g. `into` function to update deps, so that the existing deps are not lost: `(set-env! :dependencies #(into % '[[quiescent "0.3.2"] [sablono "0.7.4"]])`

Boot-cljs has it's own envinronment which is not automatically updated, but this is another problem: https://github.com/adzerk-oss/boot-cljs/issues/95

**Comment by Deraen** [comment]
Ah, I notice now that you already referenced boot-cljs issue.

> its not add to classpath current running boot-env..

It is. Boot-cljs just has it's own env and classpath, and Boot doesn't automatically update that, but the task needs to do that itself.

To check that Boot classpath includes the new dependency you can try `(io/resource "sablono/core.cljs")`  from repl and you should see that Sablono is included in root classpath.

**Comment by azizzaeny** [comment]
Ya, I see no problem for any clojure type project... We ((Lisper)) can do set-env!, even do daily life in the REPL... that's why boot is so cool.... But for clojurescript user it is not..

Boot-cljs has its own pods.. obviously its need own env to scope them from the world... but we need to update those generated `cljs_deps.js`

solution : 

 ```clojure
(def deps #(get-env :dependencies))

(defn merge-deps [& args]
  (let [dedup-deps (into [] (distinct (apply conj (seq (deps)) args)))]
    (set-env! :dependencies dedup-deps)))
```
Make easy to incrementally add deps, just eval `(merge-deps '[sablono "0.7.4"])` 

```clojure
(defonce current-deps (atom))  ; save to compare later
(def compiler-cljs (cljs))             ; maybe use atom
(boot compiler-cljs )                   ; takes long time to generate
```
To update cljs_deps,, boot task must start from scratch,..  so to fix it i use logic in middle of task....

```clojure

(defn diff-deps-tree 
  "Diff compare is any change on dependency vector"
  []) 

(defn set-compiler
 "Call cljs from scratch"
 [])

(def changed-deps? (diff-deps-tree) 

(comp 
  ... 
  (if changed-deps? 
      (set-compiler)        ;; if new one its take long
      @compiler-cljs))    ;; if reuse its take 1s

  ... )
```
Basically its do `(boot (cljs))` from scratch again.. if any dependency changed..., its tricky but work... 

afer a week this issue bother me.. now I realise this is boot-cljs issue not boot :)

**Comment by azizzaeny** [comment]
would be solved in boot-cljs 2.10...
